graviti/utility/user.py

```diff
@@ -355,3 +357,178 @@ def update(self, __m: Any = (), **kwargs: _V) -> None:  # pylint: disable=argume
+            self._keys.append(key)
+
+        self._data.__setitem__(key, value)
+
+    def __delitem__(self, key: Union[int, str]) -> None:
+        if isinstance(key, int):
+            key = self._keys.pop(key)
+        else:
+            self._keys.remove(key)
+
+        self._data.__delitem__(key)
+
+    def __contains__(self, key: Any) -> bool:
+        return self._data.__contains__(key)
+
+    def __iter__(self) -> Iterator[str]:
+        return self._keys.__iter__()
+
+    def __eq__(self, other: Any) -> bool:
+        if not isinstance(other, self.__class__):
+            return False
+
+        return self._keys.__eq__(other._keys) and self._data.__eq__(other._data)
+
+    def keys(self) -> Tuple[str, ...]:  # type: ignore[override]
+        """Get all element keys in order.
+
+        Returns:
+            A tuple containing all elements keys.
+
+        """
+        return tuple(self._keys)
+
+    def items(self) -> Tuple[Tuple[str, _V], ...]:  # type: ignore[override]
+        """Return a tuple of the ordered (key, value) pairs in dict.
+
+        Returns:
+            The ordered (key, value) pairs in dict.
+
+        """
+        return tuple((key, self._data.__getitem__(key)) for key in self._keys)
+
+    def values(self) -> Tuple[_V, ...]:  # type: ignore[override]
+        """Return a tuple of the values in dict.
+
+        Returns:
+            The values in dict.
+
+        """
+        return tuple(self._data.__getitem__(key) for key in self._keys)
+
+    def clear(self) -> None:
+        """Remove all items from the UserOrderedDict object."""
+        self._data.clear()
+        self._keys.clear()
+
+    @overload
+    def get(self, key: str) -> Optional[_V]:  # pylint: disable=arguments-differ
+        ...
+
+    @overload
+    def get(self, key: str, default: Union[_V, _T] = ...) -> Union[_V, _T]:
+        ...
+
+    def get(self, key: str, default: Any = None) -> Any:
+        """Return the value for the key if it is in the dict, else default.
+
+        Arguments:
+            key: The key for dict.
+            default: The value to be returned if key is not in the dict.
+
+        Returns:
+            The value for the key if it is in the dict, else default.
+
+        """
+        return self._data.get(key, default)
+
+    @overload
+    def pop(self, key: Union[int, str]) -> _V:  # pylint: disable=arguments-differ
+        ...
+
+    @overload
+    def pop(self, key: Union[int, str], default: Union[_V, _T] = ...) -> Union[_V, _T]:
+        ...
+
+    def pop(self, key: Union[int, str], default: Any = __marker) -> Any:
+        """Remove specified item and return the corresponding value.
+
+        Arguments:
+            key: The key for dict.
+            default: The value to be returned if the key is not in the dict and it is given.
+
+        Returns:
+            Value to be removed from the UserOrderedDict object.
+
+        """
+        if isinstance(key, int):
+            key = self._keys.pop(key)
```

Comment:
what if the `self._keys.pop` raise `IndexError`?
if the default is given, `pop` should not raise when the key does not exist.